test/blackbox-tests/test-cases/melange/missing-melc.t

```diff
@@ -0,0 +1,86 @@
+Test cases when melc is not available
+
+  $ cat > dune-project <<EOF
+  > (lang dune 3.6)
+  > (using melange 0.1)
+  > EOF
+
+  $ cat > main.ml <<EOF
+  > let () =
+  >   print_endline "hello"
+  > EOF
+
+Set up some fake environment without melc
+
+  $ mkdir _path
+  $ ln -s $(command -v dune) _path/
+  $ ln -s $(command -v ocamlc) _path/
+
+For melange.emit stanzas, an error is shown
+
+  $ cat > dune <<EOF
+  > (melange.emit
+  >  (target output)
+  >  (module_system commonjs))
+  > EOF
+
+  $ (unset INSIDE_DUNE; PATH=_path dune runtest --always-show-command-line --root .)
+  Error: A melange.emit stanza was found, but the melange compiler melc is not
+  available. Either install it by running [opam install melange], or remove the
+  melange.emit stanzas from the project.
+  [1]
+
+For libraries, if no melange.emit stanza is found, build does not fail
+
+  $ cat > dune <<EOF
+  > (library
+  >  (name lib1)
+  >  (modules :standard \ main)
+  >  (modes byte melange))
+  > (executable
+  >  (name main)
+  >  (modules main)
+  >  (modes exe byte)
+  >  (libraries lib1))
+  > EOF
+
+  $ cat > main.ml <<EOF
+  > let () =
+  >   print_endline Lib1.Lib.t
+  > EOF
+
+  $ cat > lib.ml <<EOF
+  > let t = "hello"
+  > EOF
+
+  $ (unset INSIDE_DUNE; PATH=_path dune runtest --always-show-command-line --root .)
+
+But if melange.emit stanza is found, build will fail
```

Comment:
Isn't this behavior bad? Why should we prevent the user from running an unrelated rule because melange is absent?

Reply:
Yes, but doing what you mention would require being lazier, and trying to retrieve binaries paths across at rule execution, rather than rule generation, right?

So, instead of having a rule like this:
```
...
 (action
  (chdir
   _build/default
   (run
    /home/me/code/dune/_opam/bin/melc
    -I
    .output.mobjs/melange
```

we would have an action that runs `which` and then either goes ahead and executes `melc` or shows the error.

Would that be the right direction?

Reply:
It's not needed. You can do the `which` at rule generation and delay the error until rule construction. This is precisely why `resolve_program` returns a result type that is directly usable with `Command.run`.

Reply:
I think I got what you meant. I added a change in https://github.com/ocaml/dune/pull/6403/commits/e26a0355f90be8131a187f9a771c2e4f4088e5b2 that will fail execution if the command is not found. I originally applied to `melc` only, but then thought it probable makes sense to tell users if they want to run things that are not available, so I applied the change to any command that is executed by a dune rule.

Reply:
@rgrinberg please let me know what you think about this more generic approach. 🙏